src/compiler/chainrules.jl

```diff
@@ -49,13 +49,13 @@ Convert `x` from the differentials types ChainRules uses  to the format Zygote u
 @inline wrap_chainrules_output(x) = conj(unthunk(x))  # For now we are just not going to deal with thunks
 @inline wrap_chainrules_output(x::Tuple) = map(wrap_chainrules_output, x)
 @inline wrap_chainrules_output(x::ChainRules.AbstractZero) = nothing
-@inline function wrap_chainrules_output(x::ChainRules.Composite{P, T}) where {P, T}
-  T_outer = T <: NamedTuple  ? NamedTuple : Tuple  # must be a Tuple or NamedTuple, don't care about exact parameter types
-  xp = map(wrap_chainrules_output, x)
-  convert(T_outer, xp)
+for T_outer in (:Tuple, :NamedTuple)
+  @eval @inline function wrap_chainrules_output(x::ChainRules.Composite{P, T}) where {P, T<:$T_outer}
```

Comment:
```suggestion
  # we create seperate methods rather than using a Union + a If so that we avoid a branch that changes output type
    # because nested AD on that kinda thing make Zygote less than happy.
  @eval @inline function wrap_chainrules_output(x::ChainRules.Composite{P, T}) where {P, T<:$T_outer}
```